test/bats_pipe.bats

```diff
@@ -0,0 +1,841 @@
+load test_helper
+fixtures run
+bats_require_minimum_version 1.5.0
+
+#########
+# Test Helpers
+#########
+
+describe_args() {
+  echo "Got $# args."
+
+  while [ $# -gt 0 ]; do
+    echo "arg: $1"
+    shift
+  done
+}
+
+describe_input() {
+  describe_args "$@"
+
+  local given_stdin=
+  read -rd $'\0' -t 1 given_stdin
+  if [ -n "$given_stdin" ]; then
+    echo 'stdin:'
+    local _oldIFS="$IFS"
+    IFS=$'\n'
+    # purposefully unquoted to tab multi-line output (per-line).
+    # shellcheck disable=2086
+    printf '\t%s\n' $given_stdin
+    IFS="$_oldIFS"
+  fi
+}
+
+returns_with_given_code() {
+  echo "Will return status $1."
+
+  return "$1"
+}
+
+# Necessary to prevent SIGPIPE error.
+consume_stdin_and_returns_with_given_code() {
+  echo "Will return status $1."
+
+  local unused=
+  # ignore unused.
+  # shellcheck disable=2034
+  read -rd $'\0' -t 1 unused
+
+  return "$1"
+}
+
+output_binary_data_and_returns_with_given_code() {
+  printf '\x00\xDE\xAD\xF0\x0D'
+
+  return "$1"
+}
+
+#########
+# Tests
+#########
+
+@test "run bats_pipe with no commands" {
+  run bats_pipe
```

Comment:
Some general hints for all tests here:
1. use the `-1` switch to do inline testing
2. avoid `echo $output`, there is `--print-output-on-failure` for that and if run fails due to an exit code check, it should also print `$output` (in future releases)
  - also, stderr and stdout is not differentiated in bats output, so no need to redirect 
4. (maybe) place the line count check last, to get more localized error messages first

Reply:
> use the -1 switch to do inline testing

No problem with changing `[ "$status" -eq 1 ]` to instead use `run -1 blah`.

... but I'm a little curious around intended flow and interplay with extension libs. My usage of `[ "$status" -eq 1 ]` to to mimic the functionality of [bats-assert](https://github.com/bats-core/bats-assert)'s `assert_failure` command. I happen to use bats-assert for my projects as I find it as a very nice addition to bats. Obviously, I'm not using it here as it would present a project circular dependency. However, there is this overlap of functionality with `run -N`. I'm curious about intention over that. If it's as simple as "authors could use whichever" and "`run -N` is less verbose than `[ "$status" -eq N ]`", then that's cool.

Sorry for asking and dragging out conversation. 😅

Reply:
> avoid echo $output

Done.